### PR TITLE
sync usering

### PR DESCRIPTION
**Pull request by chriku**
Currently broken:
- [x] GitHub
- [x] Refreshing tokens

### CODE REVIEW

File: github/src/main/kotlin/gropius/sync/github/GithubDataService.kt [hunk 1]
```diff
@@ -42,6 +76,44 @@ class GithubDataService(
...
+            val encodedAccountId =
+                jsonNodeMapper.jsonNodeToDeterministicString(objectMapper.valueToTree<JsonNode>(databaseId))
+            val foundImsUser =
+                imsProject.ims().value.users().firstOrNull { it.templatedFields["github_id"] == encodedAccountId }
+            if (foundImsUser != null) {
+                return foundImsUser
+            }
+        } else {
+            val foundImsUser =
+                imsProject.ims().value.users().firstOrNull { it.username == (userData?.login ?: "github") }
```
Comment (nk-coding) [inline github/src/main/kotlin/gropius/sync/github/GithubDataService.kt:96]: you use this "github" constant 5 times, maybe add a constant for this?

File: github/src/main/kotlin/gropius/sync/github/GithubDataService.kt [hunk 2]
```diff
@@ -83,11 +155,63 @@ class GithubDataService(
...
+        if (imsConfig.readUser != null) {
+            val imsUser = neoOperations.findById(imsConfig.readUser, IMSUser::class.java).awaitSingle()
+            if (imsUser.ims().value != imsProject.ims().value) {
+                TODO("Error handling")
+            }
+            userList.add(imsUser)
+        }
+        logger.info("Requesting with users: $userList")
+        return tokenManager.executeUntilWorking(imsProject.ims().value, userList) { token ->
+            val apolloClient = ApolloClient.Builder().serverUrl(URI("https://api.github.com/graphql").toString())
```
Comment (nk-coding) [inline github/src/main/kotlin/gropius/sync/github/GithubDataService.kt:182]: hardcoded GitHub API url?

File: sync/src/main/kotlin/gropius/sync/TokenManager.kt [hunk 3]
```diff
@@ -25,59 +29,77 @@ import io.ktor.serialization.kotlinx.json.*
 @ConfigurationProperties("gropius.sync")
 data class SyncConfigurationProperties(val loginServiceBase: URI, val apiSecret: String)
```
Comment (nk-coding) [inline sync/src/main/kotlin/gropius/sync/TokenManager.kt:30]: imho this deserves its own file as it is the main configuration of this module

**Review by nk-coding — CHANGES_REQUESTED** [review]

File: github/src/main/kotlin/gropius/sync/github/GithubDataService.kt (hunk 1 above)
Comment (chriku) [reply github/src/main/kotlin/gropius/sync/github/GithubDataService.kt]: 9568d5bee85f0d3491c71a29a1908da728ff092c

File: github/src/main/kotlin/gropius/sync/github/GithubDataService.kt (hunk 2 above)
Comment (chriku) [reply github/src/main/kotlin/gropius/sync/github/GithubDataService.kt]: 510952f840d5d3cb79ceb82c63bb862a6c9232bb

File: sync/src/main/kotlin/gropius/sync/TokenManager.kt (hunk 3 above)
Comment (chriku) [reply sync/src/main/kotlin/gropius/sync/TokenManager.kt]: 70a32864136ef95c728a9a3e886c287c8cdab784

File: github/src/main/kotlin/gropius/sync/github/GithubDataService.kt [hunk 4]
```diff
@@ -83,11 +159,80 @@ class GithubDataService(
...
+            userList.add(imsUser)
+        }
+        logger.info("Requesting with users: $userList")
+        return tokenManager.executeUntilWorking(imsProject.ims().value, userList) { token ->
+            val apolloClient = ApolloClient.Builder()
+                .serverUrl(imsConfig.graphQLUrl.toString())
+                .addHttpHeader("Authorization", "Bearer ${token.token}").build()
+            val res = apolloClient.mutation(body).execute()
+            logger.info("Response Code for request with token $token is ${res.data} ${res.errors}")
+            if (res.errors?.isNotEmpty() != true) Optional.of(res)
```
Comment (nk-coding) [inline github/src/main/kotlin/gropius/sync/github/GithubDataService.kt:199]: brackets pls

File: github/src/main/kotlin/gropius/sync/github/GithubDataService.kt [hunk 5]
```diff
@@ -83,11 +159,80 @@ class GithubDataService(
...
+            }
+            userList.add(imsUser)
+        }
+        logger.info("Requesting with users: $userList ")
+        return tokenManager.executeUntilWorking(imsProject.ims().value, userList) { token ->
+            val apolloClient = ApolloClient.Builder().serverUrl(imsConfig.graphQLUrl.toString())
+                .addHttpHeader("Authorization", "Bearer ${token.token}").build()
+            val res = apolloClient.query(body).execute()
+            logger.info("Response Code for request with token $token is ${res.data} ${res.errors}")
+            if (res.errors?.isNotEmpty() != true) Optional.of(res)
```
Comment (nk-coding) [inline github/src/main/kotlin/gropius/sync/github/GithubDataService.kt:234]: brackets pls

File: jira/src/main/kotlin/gropius/sync/jira/JiraDataService.kt [hunk 6]
```diff
@@ -1,33 +1,77 @@
...
+    /**
+     * Http Client for Jira API requests
+     */
+    val client = HttpClient() {
+        expectSuccess = true
+        install(Logging)
+        install(ContentNegotiation) {
+            json(Json {
+                ignoreUnknownKeys = true
+            }, contentType = ContentType.parse("application/json; charset=utf-8"))
```
Comment (nk-coding) [inline jira/src/main/kotlin/gropius/sync/jira/JiraDataService.kt:71]: `ContentType.Application.Json`?

File: jira/src/main/kotlin/gropius/sync/jira/JiraSync.kt [hunk 7]
```diff
@@ -108,31 +97,18 @@ final class JiraSync(
...
-                        appendPathSegments("comment")
-                        parameters.append("startAt", "$startAt")
-                    }
-                    headers {
-                        append(
-                            HttpHeaders.Authorization, "Basic ${basicToken}"
-                        )
-                    }
-                }.body<CommentQuery>()
+                val q = jiraDataService.request<Unit>(imsProject, listOf(), HttpMethod.Get) {
```
Comment (nk-coding) [inline jira/src/main/kotlin/gropius/sync/jira/JiraSync.kt:106]: better variable name

File: jira/src/main/kotlin/gropius/sync/jira/JiraSync.kt [hunk 8]
```diff
@@ -144,28 +120,17 @@ final class JiraSync(
...
-                    parameters.append("expand", "names,schema,editmeta,changelog")
-                    parameters.append("startAt", "$startAt")
-                }
-                headers {
-                    append(
-                        HttpHeaders.Authorization, "Basic ${basicToken}"
-                    )
-                }
-            }.body<ProjectQuery>()
+            val q = jiraDataService.request<Unit>(imsProject, listOf(), HttpMethod.Get) {
```
Comment (nk-coding) [inline jira/src/main/kotlin/gropius/sync/jira/JiraSync.kt:128]: better variable name

**Review by nk-coding — CHANGES_REQUESTED** [review]

File: jira/src/main/kotlin/gropius/sync/jira/JiraDataService.kt (hunk 6 above)
Comment (chriku) [reply jira/src/main/kotlin/gropius/sync/jira/JiraDataService.kt:71]: no. that would result in `application/json`, I need `application/json; charset=utf-8`

Comment (nk-coding) [reply jira/src/main/kotlin/gropius/sync/jira/JiraDataService.kt:71]: wtf is wrong with Jira

File: github/src/main/kotlin/gropius/sync/github/GithubDataService.kt [hunk 9]
```diff
@@ -1,37 +1,75 @@
...
 ) : SyncDataService {
 
+    companion object {
+        const val fallbackUserName = "github"
+    }
+
+    /**
+     * Logger used to print notifications
+     */
+    final val logger = LoggerFactory.getLogger(GithubDataService::class.java)
```
[the remark this answers is not in the document]
Comment (chriku) [reply github/src/main/kotlin/gropius/sync/github/GithubDataService.kt:71]: function is used from `public final inline`, I assume the logger needs most of the same

File: github/src/main/kotlin/gropius/Application.kt [hunk 10]
```diff
@@ -62,6 +62,7 @@ class Application : CommandLineRunner {
 
     override fun run(vararg args: String?) {
         try {
+            System.getProperties().stringPropertyNames().forEach(System.out::println);
```
[the remark this answers is not in the document]
Comment (chriku) [reply github/src/main/kotlin/gropius/Application.kt]: c4635d5dfd1b17a492a5081c96815f9588551ae8

File: github/src/main/kotlin/gropius/sync/github/GithubDataService.kt [hunk 11]
```diff
@@ -1,37 +1,75 @@
...
+    val labelInfoRepository: LabelInfoRepository,
+    val tokenManager: GithubTokenManager,
+    val helper: JsonHelper,
+    val objectMapper: ObjectMapper,
+    val jsonNodeMapper: JsonNodeMapper,
+    val gropiusUserRepository: GropiusUserRepository
 ) : SyncDataService {
 
+    companion object {
+        const val fallbackUserName = "github"
```
[the remark this answers is not in the document]
Comment (chriku) [reply github/src/main/kotlin/gropius/sync/github/GithubDataService.kt]: c4635d5dfd1b17a492a5081c96815f9588551ae8

File: github/src/main/kotlin/gropius/sync/github/GithubDataService.kt [hunk 12]
```diff
@@ -42,6 +80,44 @@ class GithubDataService(
         ).awaitSingle()
     }
 
+    /**
+     * Get a IMSUser for a GitHub user
+     * @param imsProject the project to map the user to
+     * @param userData the Jira user
```
[the remark this answers is not in the document]
Comment (chriku) [reply github/src/main/kotlin/gropius/sync/github/GithubDataService.kt:86]: c4635d5dfd1b17a492a5081c96815f9588551ae8

File: github/src/main/kotlin/gropius/sync/github/GithubDataService.kt (hunk 4 above)
Comment (chriku) [reply github/src/main/kotlin/gropius/sync/github/GithubDataService.kt]: c4635d5dfd1b17a492a5081c96815f9588551ae8

File: github/src/main/kotlin/gropius/sync/github/GithubDataService.kt (hunk 5 above)
Comment (chriku) [reply github/src/main/kotlin/gropius/sync/github/GithubDataService.kt]: c4635d5dfd1b17a492a5081c96815f9588551ae8

File: jira/src/main/kotlin/gropius/sync/jira/JiraSync.kt (hunk 7 above)
Comment (chriku) [reply jira/src/main/kotlin/gropius/sync/jira/JiraSync.kt]: c4635d5dfd1b17a492a5081c96815f9588551ae8

File: jira/src/main/kotlin/gropius/sync/jira/JiraSync.kt (hunk 8 above)
Comment (chriku) [reply jira/src/main/kotlin/gropius/sync/jira/JiraSync.kt]: c4635d5dfd1b17a492a5081c96815f9588551ae8

File: jira/src/main/kotlin/gropius/sync/jira/JiraDataService.kt [hunk 13]
```diff
@@ -71,12 +115,26 @@ class JiraDataService(
      * @param user the Jira user
      */
     suspend fun mapUser(imsProject: IMSProject, user: JsonElement): User {
```
Comment (nk-coding) [inline jira/src/main/kotlin/gropius/sync/jira/JiraDataService.kt:117]: missing return value doc

File: jira/src/main/kotlin/gropius/sync/jira/JiraDataService.kt [hunk 14]
```diff
@@ -87,16 +145,112 @@ class JiraDataService(
     suspend fun mapLabel(imsProject: IMSProject, label: String): Label {
```
Comment (nk-coding) [inline jira/src/main/kotlin/gropius/sync/jira/JiraDataService.kt:145]: missing return value doc

File: jira/src/main/kotlin/gropius/sync/jira/JiraDataService.kt (hunk 6 above)
Comment (nk-coding) [reply jira/src/main/kotlin/gropius/sync/jira/JiraDataService.kt:71]: but as you use it 3 times, maybe a constant?

File: sync/src/main/kotlin/gropius/sync/AbstractSync.kt [hunk 15]
```diff
@@ -107,7 +107,7 @@ abstract class AbstractSync(
      * @return Conversion information
      */
     abstract suspend fun syncComment(
-        imsProject: IMSProject, issueId: String, issueComment: IssueComment
+        imsProject: IMSProject, issueId: String, issueComment: IssueComment, users: List<User>
```
Comment (nk-coding) [inline sync/src/main/kotlin/gropius/sync/AbstractSync.kt:110]: sooo, technically you could  copy paste a users doc in all applicable methods in this file

Comment (chriku) [reply sync/src/main/kotlin/gropius/sync/AbstractSync.kt:111]: 195827e3be0eb2161f4648cee5e36ca93118f497

**Review by nk-coding — APPROVED** [review]

**Pull request merged**